src/R2N.jl

```diff
@@ -1,294 +1,512 @@
-    prox!(s, ψ, -subsolver_options.ν * ∇fk, subsolver_options.ν)
-    ξ1 = hk - mk1(s) + max(1, abs(hk)) * 10 * eps()
-    ξ1 > 0 || error("R2N: first prox-gradient step should produce a decrease but ξ1 = $(ξ1)")
-    sqrt_ξ1_νInv = sqrt(ξ1 * νInv)
+  mk1 = let ψ = ψ
+            d -> φ1(d) + ψ(d)::T
+  end
 
-    if ξ1 ≥ 0 && k == 1
-      ϵ_increment = ϵr * sqrt_ξ1_νInv
-      ϵ += ϵ_increment  # make stopping test absolute and relative
-      ϵ_subsolver += ϵ_increment
-    end
+  mk = let ψ = ψ, solver = solver
+           d -> obj(solver.subpb.model, d) + ψ(d)::T
+  end
 
-    if sqrt_ξ1_νInv < ϵ
-      # the current xk is approximately first-order stationary
-      optimal = true
-      continue
-    end
-    s1 = copy(s)
-
-    subsolver_options.ϵa = k == 1 ? 1.0e-3 : min(sqrt_ξ1_νInv^(1.5), sqrt_ξ1_νInv * 1e-3)
-    verbose > 0 && @debug "setting inner stopping tolerance to" subsolver_options.optTol
-    subsolver_options.σk = σk
-    subsolver_args = subsolver == R2DH ? (SpectralGradient(νInv, f.meta.nvar),) : ()
-    s, iter, _ = with_logger(subsolver_logger) do
-      subsolver(φ, ∇φ!, ψ, subsolver_args..., subsolver_options, s)
-    end
+  prox!(s1, ψ, mν∇fk, ν₁)
+  mks = mk1(s1)
+
+  ξ1 = hk - mks + max(1, abs(hk)) * 10 * eps()
+  sqrt_ξ1_νInv = ξ1 ≥ 0 ? sqrt(ξ1 / ν₁) : sqrt(-ξ1 / ν₁)
+  solved = (ξ1 < 0 && sqrt_ξ1_νInv ≤ neg_tol) || (ξ1 ≥ 0 && sqrt_ξ1_νInv ≤ atol)
+  (ξ1 < 0 && sqrt_ξ1_νInv > neg_tol) &&
+    error("R2N: prox-gradient step should produce a decrease but ξ1 = $(ξ1)")
+  atol += rtol * sqrt_ξ1_νInv # make stopping test absolute and relative
+  
+  set_solver_specific!(stats, :xi, sqrt_ξ1_νInv)
+  set_status!(
+    stats,
+    get_status(
+      reg_nlp,
+      elapsed_time = stats.elapsed_time,
+      iter = stats.iter,
+      optimal = solved,
+      improper = improper,
+      max_eval = max_eval,
+      max_time = max_time,
+      max_iter = max_iter,
+    ),
+  )
+
+  callback(nlp, solver, stats)
+
+  done = stats.status != :unknown
+
+  while !done
+
+    sub_atol = stats.iter == 0 ? 1.0e-3 : min(sqrt_ξ1_νInv ^ (1.5) , sqrt_ξ1_νInv * 1e-3)
+    
+    solver.subpb.model.σ = σk
+    isa(solver.subsolver, R2DHSolver) && (solver.subsolver.D.d[1] = 1/ν₁)
+    ν_sub = isa(solver.subsolver, R2DHSolver) ? 1/σk : ν₁
+    solve!(
+      solver.subsolver, 
+      solver.subpb, 
+      solver.substats;
+      x = s1,
+      atol = sub_atol,
+      ν = ν_sub,
+      kwargs...
+      )
+
+    s .= solver.substats.solution
 
     if norm(s) > β * norm(s1)
       s .= s1
     end
 
-    # restore initial subsolver_options.ϵa here so that subsolver_options.ϵa
-    # is not modified if there is an error
-    subsolver_options.ν = ν_subsolver
-    subsolver_options.ϵa = ϵ_subsolver_init
-    subsolver_options.σk = σk
-    Complex_hist[k] = iter
-
     xkn .= xk .+ s
-    fkn = obj(f, xkn)
-    hkn = h(xkn[selected])
+    fkn = obj(nlp, xkn)
+    hkn = @views h(xkn[selected])
     hkn == -Inf && error("nonsmooth term is not proper")
```

Comment:
This could happen if $\sigma$ is not large enough, but doesn't mean $h$ is not proper. If `hkn == -Inf`, the step will be rejected and `σk` will be increased.